src/utils/IdealGraphVisualizer/HierarchicalLayout/src/main/java/com/sun/hotspot/igv/hierarchicallayout/HierarchicalStableLayoutManager.java

```diff
@@ -0,0 +1,1909 @@
+            return n.x;
+        }
+
+        int[] values = new int[n.preds.size() + n.succs.size()];
+        int i = 0;
+
+        for (LayoutEdge e : n.preds) {
+            values[i] = e.from.x + e.relativeFrom - e.relativeTo;
+            i++;
+        }
+
+        for (LayoutEdge e : n.succs) {
+            values[i] = e.to.x + e.relativeTo - e.relativeFrom;
+            i++;
+        }
+
+        return median(values);
+    }
+
+    private int median(int[] values) {
+        Arrays.sort(values);
+        if (values.length % 2 == 0) {
+            return (values[values.length / 2 - 1] + values[values.length / 2]) / 2;
+        } else {
+            return values[values.length / 2];
+        }
+    }
+
+    /**
+     * Adjust the X-coordinates of the nodes in the given layer, as a new node has
+     * been inserted at that layer
+     *
+     * @param layer
+     */
+    private void adjustXCoordinates(int layer) {
+        List<LayoutNode> nodes = layers.get(layer);
+        ArrayList<Integer> space = new ArrayList<>();
+        List<LayoutNode> nodeProcessingOrder = new ArrayList<>();
+
+        nodes.sort(nodePositionComparator);
+
+        int curX = 0;
+        for (LayoutNode n : nodes) {
+            space.add(curX);
+            curX += n.width + X_OFFSET;
+            nodeProcessingOrder.add(n);
+        }
+
+        nodeProcessingOrder.sort(nodeProcessingUpComparator);
+        NodeRow r = new NodeRow(space);
+        for (LayoutNode n : nodeProcessingOrder) {
+            int optimal = calculateOptimalBoth(n);
+            r.insert(n, optimal);
+        }
+    }
+
+    /**
+     * Ensure that the datastructures nodes and layerNodes are consistent
+     */
+    private void sanityCheckNodesAndLayerNodes() {
+        int nodeCount = 0;
+        for (int i = 0; i < layers.keySet().size(); i++) {
+            assert layers.containsKey(i);
+            layers.get(i).sort(nodePositionComparator);
+            int nodePos = 0;
+            for (LayoutNode n : layers.get(i)) {
+                assert n.layer == i;
+                assert nodes.contains(n);
+                assert n.pos == nodePos;
+                nodePos += 1;
+                nodeCount += 1;
+            }
+        }
+        for (LayoutNode n : nodes) {
+            assert n.vertex == null || vertexToLayoutNode.get(n.vertex).equals(n);
+            assert n.layer < layers.keySet().size();
+            assert layers.get(n.layer).contains(n);
+        }
+        assert nodeCount == nodes.size();
+    }
+
+    private void sanityCheckEdges() {
+        for (LayoutNode n : nodes) {
+            for (LayoutEdge e : List.copyOf(n.succs)) {
+                assert e.from.equals(n);
+                if (nodes.contains(e.to)) {
+                    assert e.to.preds.contains(e);
+                    assert layers.get(n.layer + 1).contains(e.to);
+                    assert e.to.layer == n.layer + 1;
+                } else {
+                    n.succs.remove(e);
```

Comment:
Yes, it breaks often if removed. It's mostly sanity checking but also ensuring the state of the graph is correct by fixing weird edge connections

Reply:
As mentioned below, this method is still a bottleneck and we spend quite some time inside it. Overall, it attributes to around 90% of the overall time to open the graph in the example mentioned earlier. Is there another way to fix these breakages instead of looping over all nodes and edges here?

Reply:
I am not sure, I've tried to find other solutions but have not managed. The function can be called less, but not removed entirely. A suggestion is to put it up as a future enhancement?